LibELF+LibC+Kernel/aarch64: Implement everything needed to dynamically load an executable

FireFox317 opened the pull request:
~~**NOTE: This PR is currently a draft as it needs to be applied on top of #17249. So the first 20 commits until 99196f7bee788dece90a7e61084dab4deebfaa5e can be ignored!**~~

This PR contains the necessary patches to LibELF, LibC and the Kernel to be able to successfully jump to the entry point of SystemServer. This means the DynamicLoader is completely implemented for aarch64. To get there the main changes of this PR are the following:

- Implement TLS (Thread Local Storage) for aarch64 in bf6fbf7d7833349889e4582f9b3660c2eac2a607
- Handle all the aarch64 relocation types. The code for this could be shared with x86_64, which makes the implementation very straight forward (f7e216560b0a3af14e5a5d5e1ab82624362ca2c6 and ae28d3affae2999f59566b3185922cd2445576f2)
- Handle SVC instructions used for syscalls in cf16ba54991da2ea7752ae0b9c5791bafa2efe7e
- Other various smaller changes

With this code in place, we now execute a `fork` in `SystemServer`, which hits a TODO in `sys$fork`.
![image](https://user-images.githubusercontent.com/18349222/215761130-42634a9e-70de-4c4f-b00a-a136dc3fbf39.png)

nico commented on Userland/DynamicLoader/CMakeLists.txt:
```diff
@@ -30,6 +30,11 @@ set(SOURCES ${LOADER_SOURCES} ${AK_SOURCES} ${ELF_SOURCES} ${LIBC_SOURCES1} ${LI
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-rtti -nostdlib -pie -fpic -DNO_TLS")
 
+if ("${SERENITY_ARCH}" STREQUAL "aarch64")
+    # On aarch64 the stack protector will be accessed before the Loader can relocate itself..
```
nits: s/will/would/, and one period at the end is enough.

Also the first `,` in the commit description should be omitted I think.

FireFox317 replied:
Fixed!

nico commented on Kernel/Arch/aarch64/Exceptions.cpp:
```diff
@@ -78,6 +78,13 @@ static void setup_el1()
 
+    Aarch64::CPACR_EL1 cpacr_el1 = {};
+    cpacr_el1.ZEN = 0;     // Trap SVE instructions at EL1 and EL0
+    cpacr_el1.FPEN = 0b11; // Don't trap Advanced SIMD and floating-point instructions
```
The kernel shouldn't use any floating point instructions (…else we'd have to save the float registers on context switch), right? Why do we need this?

FireFox317 replied:
We need this because userspace (EL0) will execute floating-point instructions. According to the docs (https://developer.arm.com/documentation/ddi0595/2021-03/AArch64-Registers/CPACR-EL1--Architectural-Feature-Access-Control-Register?lang=en#fieldset_0-21_20) there is no option to only trap on floating-point instruction in EL1, it's only possible to completely disable the trapping for both EL0 and EL1.

I went ahead and implemented storing/restoring of the floating point registers for the userspace threads, because we need that anyway.

nico commented on Kernel/Arch/aarch64/Exceptions.cpp:
```diff
@@ -74,7 +74,10 @@ static void setup_el1()
+
+    // NOTE: Disable memory access alignment check as we cannot enforce
+    //       proper alignment for userspace applications.
+    system_control_register_el1.A = 0;
```
Why can't we require userspace apps to only do aligned memory accesses?

Do the currently-loaded first-party userspace apps do unaligned reads? Where?

FireFox317 replied:
We could require that by adding the same flags as we do for the kernel binary (`-mstrict-align -Wno-cast-align`), however that means we also need to add those flags for all the ports. I thought it was better if we don't enforce that for userspace applications. And it's not possible to only allow unaligned accesses for EL0, so we need to change that in SCTLR_EL1.

It happens all over the place in our libraries, I don't remember where exactly I hit it tho.

FireFox317 replied:
I'm fine with either approach, but since we discussed and enabled the enforcement of aligned memory accesses for the kernel, I think it's okay to do the same for userspace now. So I went ahead and removed the previous commit that would disable the check, and instead added the `-mstrict-align -Wno-cast-align` flags to `serenity_compile_options.cmake` in 86b0978dd07653cff761ec57ab25f5985cc7ab8f

nico commented on Userland/Libraries/LibC/string.cpp:
```diff
@@ -136,29 +136,26 @@ void* memcpy(void* dest_ptr, void const* src_ptr, size_t n)
+    u8 const* ps = (u8 const*)src_ptr;
+    for (; n--;)
+        *pd++ = *ps++;
+    return dest_ptr;
```
I wonder if compilers are smart enough to see that this is memcpy(), then transform it to a call to `__builtin_memset()` internally, which can then generate a call to memcpy() under certain circumstances. Do we use any flags or labels.

Well, if it becomes a problem, we can compile this file with `-fno-builtin-memset -fno-builtin-memcpy` I suppose.

(Or maybe my mental model on how this works is incomplete – well possible.)

FireFox317 replied:
We add these flags for the LibC build, so I think it should be fine (at least for the GCC build).
https://github.com/SerenityOS/serenity/blob/84e8d5f41842f6c3f9f613f3570f669328fe4d06/Userland/Libraries/LibC/CMakeLists.txt#L138-L141

nico commented on Userland/Libraries/LibELF/DynamicLoader.cpp:
```diff
@@ -482,7 +482,8 @@ DynamicLoader::RelocationResult DynamicLoader::do_relocation(const ELF::DynamicO
         break;
-    case R_X86_64_64: {
+    case R_X86_64_64:
+    case R_AARCH64_ABS64: {
```
very nit: I'd put the AARCH64 case label first, so that the case labels are alphabetically ordered. (also below.)

FireFox317 replied:
Fixed!

nico reviewed (commented):
Looks great, and a lot less code than I though from the commit descriptions :)

Two real questions below, the rest are nits. (The real questions are:

1. Kernel FPU why
2. Unaligned mem

)

FireFox317 commented:
Thanks again for the review Nico! I went ahead and implemented storing and restoring the floating-point registers, as this PR also includes the commit that disables the trapping of the floating-point instructions.

nico commented:
Sorry for missing this!

For the FPU stuff, my question was: Kernels usually don't use FPU instructions themselves afaik, so that kernel<->user context switches don't need to store FPU registers.

Is the code you added here for user<->user context switches and for user code only?

FireFox317 commented:
> Sorry for missing this!

No worries! I updated the code according to your comments!


> Is the code you added here for user<->user context switches and for user code only?

Right, the code that I added will only store and restore the FPU state when the `Processor::switch_context` function is called. That function is called from the Scheduler, and will cause Threads to be switched, and the code currently doesn't distinguish kernel vs user thread. The x86_64 code also does not take that into account.

However note that when a userland process enters the kernel via a syscall, we do not call the `switch_context` function, so we also don't store/restore FPU registers at that point. It only happens when the Scheduler decides to change to a different running thread.

nico reviewed (approved):
LG!

FireFox317 commented:
Turns out the CI failing was an actual problem, should be fixed now :)

ADKaster commented on Userland/Libraries/LibELF/Arch/aarch64/entry.S:
```diff
@@ -8,5 +9,4 @@
 .type _invoke_entry,@function
 _invoke_entry: # (argc, argv, envp, entry)
-    # FIXME: Possibly incomplete.
-    ret
```
Iirc the x86_64 implementation here takes care to ensure alignment of the stack matches the SysV ABI (16 bytes). Do we need a similar guarantee for aarch64?

FireFox317 replied:
The aarch64 hardware will actually trap if the stack pointer is not 16-byte aligned, and this code works as expected. So that means that hardware actually enforces it, and an exception will be generated once it is not the case. So I think this is fine, and if in the future the codegen of the compiler change, such that it is not aligned anymore, we'll be notified by that because of the hardware enforcement.
